File: benchmarks/54-eigenproblems/eigenproblems.py
#!/usr/bin/env python3
"""Experiment 54 -- Eigenproblems (scipy column).

Same kernels as ``eigenproblems.m``, same order and labels.

EXECUTION comparison against compiled LAPACK (scipy.linalg) and ARPACK
(scipy.sparse.linalg).  Timing on large random matrices; checks on the small
reproducible spd_int_matrix(8) or a fixed matrix, as order-invariant scalars.

Two Mathilda gaps this measures: Eigensystem is absent (the .m side SKIPs it),
and the generalized problem has no LAPACK path there -- here scipy.linalg.eigh
does it in microseconds, so the 5x5 loop shows the gap.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import math
import numpy as np
import scipy.linalg as sla
import scipy.sparse.linalg as ssla

from harness import bench, check, require, seed
from data import spd_int_matrix

require(["numpy.linalg:eigvals", "numpy.linalg:eigvalsh", "scipy.linalg:eigh",
         "scipy.linalg:eig", "scipy.sparse.linalg:eigsh"])


def r(x, p):
    return int(math.floor(float(x) * 10 ** p + 0.5))


seed()
ns = 300
m = np.random.random((ns, ns)); sy = (m + m.T) / 2               # symmetric
ge = np.random.random((ns, ns))                                  # general
mv = np.random.random((250, 250)); sv = (mv + mv.T) / 2          # eigenvectors
mb = np.random.random((500, 500)); big = (mb + mb.T) / 2         # Arnoldi

sm = spd_int_matrix(8)
gen4 = np.array([[3, 1, 0, 2], [2, 4, 1, 0], [0, 1, 5, 1], [1, 0, 1, 6]], dtype=float)
A5 = spd_int_matrix(5)
B5 = np.diag([2., 3., 4., 5., 6.])

bench("Eigenvalues symmetric 300x300", lambda: np.linalg.eigvalsh(sy), reps=1)
check("Eigenvalues symmetric 300x300", r(np.linalg.eigvalsh(sm).max(), 6))

bench("Eigenvalues general 300x300", lambda: np.linalg.eigvals(ge), reps=1)
check("Eigenvalues general 300x300", r(np.abs(np.linalg.eigvals(gen4)).sum(), 4))

bench("Eigenvectors symmetric 250x250", lambda: np.linalg.eigh(sv), reps=1)  # no check

# Generalized symmetric-definite eigenvalues -- scipy.linalg.eigh(A, B), one LAPACK
# call, vs Mathilda's symbolic char-poly path. Same 5x5 loop shape.
bench("Generalized eigenvalues 5x5 x50", lambda: [sla.eigh(A5, B5) for _ in range(50)], reps=1)
check("Generalized eigenvalues 5x5 x50", r(sla.eigh(A5, B5, eigvals_only=True).sum(), 6))

bench("Eigenvalues Arnoldi k=6 of 500x500", lambda: ssla.eigsh(big, k=6))
check("Eigenvalues Arnoldi k=6 of 500x500", r(np.linalg.eigvalsh(sm).sum(), 6))

# Eigensystem is ABSENT in Mathilda -> .m SKIPs; scipy returns (w, v) together.
bench("Eigensystem symmetric 250x250", lambda: np.linalg.eigh(sv), reps=1)
check("Eigensystem symmetric 250x250", r(np.linalg.eigvalsh(sm).max(), 6))

File: benchmarks/54-eigenproblems/eigenproblems.m
(* ==========================================================================
   Experiment 54 -- Eigenproblems (group E, advanced numerical analysis)
   ==========================================================================
   WHAT IT MEASURES.  The eigensolvers in src/linalg/eigen_*.c -- the symmetric
   (syev) and general (geev) dense drivers, the generalized problem
   Eigenvalues[{A, B}], and the Krylov path Method -> "Arnoldi".  Group B times
   only Eigenvalues on a single dense matrix; this separates the algorithms.

   TIMING on large random matrices; CHECKS on the small reproducible
   spdIntMatrix[8] or a fixed matrix.  Eigen-outputs are non-unique (order, sign,
   eigenvector scale), so checks are order-invariant scalars: max eigenvalue,
   sum of |eigenvalues|, trace via eigenvalues.  Eigenvectors have no stable
   scalar (Mathematica does not normalise them), so that case carries no check.

   ONE FINDING IS BAKED IN.  Eigensystem is ABSENT (benchIf).

   The generalized problem Eigenvalues[{A, B}] now takes a LAPACK machine path
   (dsygv for the symmetric-definite pencil here, dggev/zhegv/zggev otherwise;
   eigen_direct.c), so the 5x5 loop runs in sub-millisecond -- FASTER than
   scipy.linalg.eigh -- instead of forming Det[A - lambda B] and symbolically
   root-finding it (Root[] objects by size 6, no scaling past ~7).  The small
   5x5 loop is kept so the row stays comparable across the history.
   ========================================================================== *)

Get["../harness.m"];
Get["../data.m"];

require[{"Eigenvalues", "Eigenvectors", "Eigensystem"}];

seed[];
ns = 300;
sy  = With[{a = rand01[{ns, ns}]}, (a + Transpose[a])/2];       (* symmetric *)
ge  = rand01[{ns, ns}];                                          (* general *)
sv  = With[{a = rand01[{250, 250}]}, (a + Transpose[a])/2];     (* eigenvectors *)
big = With[{a = rand01[{500, 500}]}, (a + Transpose[a])/2];     (* Arnoldi *)

sm   = N[spdIntMatrix[8]];
gen4 = N[{{3, 1, 0, 2}, {2, 4, 1, 0}, {0, 1, 5, 1}, {1, 0, 1, 6}}];
A5   = N[spdIntMatrix[5]];
B5   = N[DiagonalMatrix[{2, 3, 4, 5, 6}]];

bench["Eigenvalues symmetric 300x300", Eigenvalues[sy];, 1];
check["Eigenvalues symmetric 300x300", Round[10^6 Max[Eigenvalues[sm]]]];

bench["Eigenvalues general 300x300", Eigenvalues[ge];, 1];
check["Eigenvalues general 300x300", Round[10^4 Total[Abs[Eigenvalues[gen4]]]]];

bench["Eigenvectors symmetric 250x250", Eigenvectors[sv];, 1];  (* no check: eigenvectors non-unique *)

(* Generalized: no machine path -- symbolic char-poly root-finding.  Small 5x5
   loop so it completes; the ratio against scipy.linalg.eigh(A,B) is the story. *)
bench["Generalized eigenvalues 5x5 x50", Do[Eigenvalues[{A5, B5}], {50}];, 1];
check["Generalized eigenvalues 5x5 x50", Round[10^6 Total[Eigenvalues[{A5, B5}]]]];

bench["Eigenvalues Arnoldi k=6 of 500x500", Eigenvalues[big, 6, Method -> "Arnoldi"];];
check["Eigenvalues Arnoldi k=6 of 500x500", Round[10^6 Total[Eigenvalues[sm]]]];

benchIf["Eigensystem symmetric 250x250", "Eigensystem", Eigensystem[sv];, 1];
checkIf["Eigensystem symmetric 250x250", "Eigensystem", Round[10^6 Max[Eigenvalues[sm]]]];
